src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -845,18 +847,260 @@ public Document GetRevisionBefore(DocumentsOperationContext context, string id,
+                        wasConflicted = true;
+                    }
+                }
+
                 return result;
             }
+        }
+
+
+        public class RevertProgress : IOperationProgress
+        {
+            public int ScannedRevisions;
+            public int RevertedDocuments;
+            public int ScannedDocuments;
+            public Dictionary<string, string> Warnings = new Dictionary<string, string>();
+
+            public void Warn(string id, string message)
+            {
+                Warnings[id] = message;
+            }
+
+            public DynamicJsonValue ToJson()
+            {
+                return new DynamicJsonValue
+                {
+                    [nameof(ScannedRevisions)] = ScannedRevisions,
+                    [nameof(ScannedDocuments)] = ScannedDocuments,
+                    [nameof(RevertedDocuments)] = RevertedDocuments,
+                    [nameof(Warnings)] = DynamicJsonValue.Convert(Warnings)
+                };
+            }
+        }
+
+        public class RevertResult : IOperationResult
+        {
+            public RevertProgress Progress = new RevertProgress();
+            public string Message { get; }
+            public DynamicJsonValue ToJson()
+            {
+                return new DynamicJsonValue
+                {
+                    [nameof(Progress)] = Progress.ToJson(),
+                    [nameof(Message)] = Message,
+                    [nameof(ShouldPersist)] = ShouldPersist
+                };
+            }
+
+            public bool ShouldPersist { get; }
+        }
+
+        private const long SizeLimit = 32 * 1_024 * 1_024;
+
+        public IOperationResult RevertRevisions(DateTime before, TimeSpan window, Action<IOperationProgress> onProgressAction = null)
+        {
+            var dateToBreak = before.Add(-window);
+
+            var list = new List<Document>();
+            var scannedIds = new HashSet<string>();
+            var result = new RevertResult();
+
+            var hasMore = true;
+            while (hasMore)
+            {
+                using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext writeCtx))
+                {
+                    hasMore = PrepareRevertedRevisions(writeCtx, before, dateToBreak ,result, list, scannedIds);
+                    WriteRevertedRevisions(list, writeCtx);
+                    onProgressAction?.Invoke(result.Progress);
+                    result.Progress.Warnings.Clear();
+                }
+            }
+
+            return result;
+        }
+
+        private void WriteRevertedRevisions(List<Document> list, DocumentsOperationContext context)
+        {
+            if (list.Count > 0)
+            {
+                using (var tx = context.OpenWriteTransaction())
+                {
+                    foreach (var document in list)
```

Comment:
This can be a _very_ big list.
Need to portion it off to several independent operations so as not to block the transaction merger completely.